### PR TITLE
Fix missing key issue in member access

### LINKED ISSUE #30707: ballerina: Oh no, something really went wrong. Bad. Sad.
I'm working through the examples for creating modules (https://ballerina.io/learn/user-guide/structuring-ballerina-code/working-with-ballerina-modules/). The main.bal file I entered by hand looked like so:

```
import ballerina/io;
import powelldean/winery.storage;
import winery.model;

public function main() {
    model.Wine[] wines = storage.getWineList();
    foreach var wine in wines {
      io:println(wine);
    }
}
```

Of course, I entered two typos on line 6. The periods (.) should be colons(:)
- model.Wine[] should actually be model:Wine[]
- storage.getWineList() should actually be storage:getWineList()

As soon as I fixed the typos, the sample program worked fine. The compiler should flag the syntax error, not crash.

The Ballerina-internal.log is below:

```
[2021-05-21 00:08:30,549] SEVERE {b7a.log.crash} - Index: '0', Size: '0' 
java.lang.IndexOutOfBoundsException: Index: '0', Size: '0'
	at io.ballerina.compiler.internal.syntax.NodeListUtils.rangeCheck(NodeListUtils.java:38)
	at io.ballerina.compiler.syntax.tree.SeparatedNodeList.get(SeparatedNodeList.java:45)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2109)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:475)
	at io.ballerina.compiler.syntax.tree.IndexedExpressionNode.apply(IndexedExpressionNode.java:58)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:4981)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4952)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2821)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:475)
	at io.ballerina.compiler.syntax.tree.ExpressionStatementNode.apply(ExpressionStatementNode.java:50)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4853)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1458)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:475)
	at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1365)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1345)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1323)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:475)
	at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:592)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:475)
	at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:505)
	at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:113)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:365)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:329)
	at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:176)
	at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:168)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:97)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:206)
	at io.ballerina.projects.Package.getCompilation(Package.java:131)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:68)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:272)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

Many thanks.

Dean Powell
Edmonton, Canada

### PR DESCRIPTION
**Pull request by SandaruJayawardana**
Fixes #30707

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/member_access_negative.bal [hunk 1]
```diff
@@ -196,9 +196,16 @@ function testInvalidRecordMemberAccessByConst() {
...
+};
+
 function testMemberAccessWithoutIndex() {
     int[] intArray = [];
-    int value = intArray[];
+    Abc a = {s: ["1.0", "2.0", "3.0"], v: [1.0, 2.0, 3.0]};
+    a.s[] = "4.1";
+    a.v[1] = intArray[];
     intArray[]
 }
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/member_access_negative.bal:210]: Shall we write a test case for the exact two cases mentioned in the issue?

Comment (SandaruJayawardana) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/access/member_access_negative.bal:210]: Okay. I'll add a test.

**Review by lochana-chathura — COMMENTED** [review]
@rdulmina there seem to be several places we create `IndexedExpressionNode` in the parser. Also, they are invoked through several paths. So it looks okay for me to handle the missing key at the BLangNodeTransformer in case it is not handled in the parser itself. WDYT?

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/module-test-negative.bal [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/module-test-negative.bal:1]: Change the copyright year to 2021. This is a new file

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/modules/test.shapes/test.shapes.bal [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/modules/test.shapes/test.shapes.bal:1]: Same here.. copyright year

File: tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/Ballerina.toml [hunk 4]
```diff
@@ -0,0 +1,4 @@
+[package]
```
Comment (gimantha) [inline tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/Ballerina.toml:1]: This change is not related to JVM stuff. This is a parser related test, Shall we move this project out from jvm package? and move to a place where we keep parser related tests?

Comment (gimantha) [reply tests/jballerina-unit-test/src/test/resources/test-src/jvm/TestProjectNegative/Ballerina.toml:1]: `parser` module seems good

**Comment by rdulmina** [comment]
> @rdulmina there seem to be several places we create `IndexedExpressionNode` in the parser. Also, they are invoked through several paths. So it looks okay for me to handle the missing key at the BLangNodeTransformer in case it is not handled in the parser itself. WDYT?

Since the parser should always return a valid syntax tree I don't think it is the correct way. I think we should fix this from the parser

**Comment by lochana-chathura** [comment]
> Since the parser should always return a valid syntax tree I don't think it is the correct way. I think we should fix this from the parser

As per our offline discussion we agreed upon handling this in both the parser(for the known scenarios) and BlangNodetranformer(as a contingency plan). 

@SandaruJayawardana shall we do that?

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -18179,6 +18180,21 @@ private STNode getExpression(STNode ambiguousNode) {
                         errorCons.typeReference, errorCons.openParenToken, errorArgs, errorCons.closeParenToken);
             case IDENTIFIER_TOKEN:
                 return STNodeFactory.createSimpleNameReferenceNode(ambiguousNode);
+            case INDEXED_EXPRESSION:
+                STNodeList keys = (STNodeList) ((STIndexedExpressionNode) ambiguousNode).keyExpression;
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:18184]: Since `(STIndexedExpressionNode) ambiguousNode` is repeated, shall we extract it to a separate variable?

**Review by lochana-chathura — COMMENTED** [review]
LGTM except for small refactoring.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java (hunk 5 above)
Comment (SandaruJayawardana) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:18184]: I'll change it.

**Review by lochana-chathura — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 6]
```diff
@@ -154,6 +154,8 @@
             "BCE2115", "required.param.not.allowed.after.included.record.param"),
     INCOMPATIBLE_SUB_TYPE_FIELD("BCE2116", "incompatible.sub.type.field"),
 
+    ERROR_MISSING_KEY_EXPR_IN_MEMBER_ACCESS_EXPR("BCE2117", "missing.key.expr.in.member.access.expr"),
```
Comment (rdulmina) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:157]:
```suggestion
    MISSING_KEY_EXPR_IN_MEMBER_ACCESS_EXPR("BCE2117", "missing.key.expr.in.member.access.expr"),
```

Comment (rdulmina) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:157]: Let's remove the newline at the beginning as well

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/AccessExpressionsTest.java [hunk 7]
```diff
@@ -167,4 +167,9 @@ public void testOptionalFieldAccessWithMissingExpression() {
     public void testMemberAccessRecovery() {
         testFile("access-expr/member_access_expr_source_11.bal", "access-expr/member_access_expr_assert_11.json");
     }
+
+    @Test
+    public void testModuleCallSyntaxErrors() {
+        testFile("access-expr/module_call_source.bal", "access-expr/module_call_assert.json");
+    }
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/AccessExpressionsTest.java:174]: Let's move this test to array type desc tests and change the names accordingly

Comment (rdulmina) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/AccessExpressionsTest.java:174]: bal file name -> `array_type_source_21.bal`
source file name -> `array_type_assert_21.json`
testName -> `testInvalidArrayBaseExpression`

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/AccessExpressionsTest.java:174]: +1

Comment (SandaruJayawardana) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/expressions/AccessExpressionsTest.java:174]: Okay. I'll move.

**Review by rdulmina — APPROVED** [review]

**Review by lochana-chathura — APPROVED** [review]
